packages/excalidraw/change.ts

```diff
@@ -0,0 +1,1273 @@
+      }
+    } else if ((boundText as ExcalidrawTextElement).containerId) {
+      // Unbind when we cannot find the container
+      return newElementWith(boundText as Ordered<ExcalidrawTextElement>, {
+        containerId: null,
+      });
+    }
+  }
+
+  private static redrawTextBoundingBoxes(
+    changed: ReadonlyMap<string, OrderedExcalidrawElement>,
+    elements: SceneElementsMap,
+  ) {
+    const boxesToRedraw = new Map<
+      string,
+      { container: OrderedExcalidrawElement; boundText: ExcalidrawTextElement }
+    >();
+
+    for (const element of changed.values()) {
+      if (ElementsChange.whenBoundText(element)) {
+        const { containerId } = element as ExcalidrawTextElement;
+        const container = containerId ? elements.get(containerId) : undefined;
+
+        if (container) {
+          boxesToRedraw.set(container.id, {
+            container,
+            boundText: element as ExcalidrawTextElement,
+          });
+        }
+      }
+
+      if (ElementsChange.whenTextContainer(element)) {
+        const boundTextElementId = getBoundTextElementId(element);
+        const boundText = boundTextElementId
+          ? elements.get(boundTextElementId)
+          : undefined;
+
+        if (boundText) {
+          boxesToRedraw.set(element.id, {
+            container: element,
+            boundText: boundText as ExcalidrawTextElement,
+          });
+        }
+      }
+    }
+
+    for (const { container, boundText } of boxesToRedraw.values()) {
+      if (container.isDeleted || boundText.isDeleted) {
+        // Skip on deleted container or bound text, as it doesn't result in visible changes so we don't need to redraw
+        continue;
+      }
+
+      // TODO: refactor mutations away, so we couldn't end up in an incosistent state
+      redrawTextBoundingBox(boundText, container, elements, false);
+    }
+  }
+
+  private static reorderElements(
+    elements: SceneElementsMap,
+    flags: {
+      containsVisibleDifference: boolean;
+      containsZindexDifference: boolean;
+    },
+  ) {
+    if (!flags.containsZindexDifference) {
+      return elements;
+    }
+
+    const previous = Array.from(elements.values());
+    const reordered = orderByFractionalIndex([...previous]);
+
+    if (
+      !flags.containsVisibleDifference &&
+      Delta.isRightDifferent(previous, reordered, true)
+    ) {
+      // We found a difference in order!
+      flags.containsVisibleDifference = true;
+    }
+
+    return arrayToMap(reordered) as typeof elements;
+  }
+
+  /**
+   * It is necessary to post process the partials in case of reference values,
+   * for which we need to calculate the real diff between `deleted` and `inserted`.
+   */
+  private static postProcess(
+    deleted: ElementPartial,
+    inserted: ElementPartial,
+  ): [ElementPartial, ElementPartial] {
+    if (deleted.boundElements && inserted.boundElements) {
+      const deletedDifferences = arrayToObject(
```

Comment:
I rechecked reference props of other elements and it seems that we don't need have do this kind of merging (not to override other clients changes) in other cases, at least not now.

Closest next thing might be merging of linear element points - in case multiple clients modify the same elements and later undo / redo is perfomed. ATM one client will completely override the other clients changes, but it's even questionable if we should perform merge in such case as it might lead to broken linear element when we would merge the changes.

In the future, we might consider a special runtime brand or a decorator, which we would put over a given instance and it would automatically postprocess it (getting just the changes) & automatically merge the changes when applying it.